packages/search/src/use-fuzzy-search.ts

```diff
@@ -0,0 +1,69 @@
+import Fuse from 'fuse.js';
+import { useState, useMemo, useEffect } from 'react';
+
+const defaultOptions = {
+	threshold: 0.4,
+	distance: 20,
+};
+
+type FieldsProp< T > = T extends string
+	? {
+			fields?: never;
+	  }
+	: {
+			fields: Fuse.IFuseOptions< T >[ 'keys' ];
+	  };
+
+export type UseFuzzySearchOptions< T > = {
+	data: T[];
+	options?: Partial< Fuse.IFuseOptions< T > >;
```

Comment:
Out of curiosity (I'm new to TypeScript), what does `< Fuse.IFuseOptions< T > >` designate?

Reply:
`IFuseOptions< T >` is the type the `Fuse` class constructor takes as a second argument. https://github.com/krisk/Fuse/blob/7ff9c2f2ea8345287b05f7560799c6fee92325e3/src/index.d.ts#L7

`T` here represents the data type that we're searching. Since the Fuse library supports searching arbitrary data types, it means we have to refer to it as an arbitrary type (a.k.a. a "generic" type). `IFuseOptions` needs to know what that generic type is though because it changes what the options object looks like, that's why we pass the `T` type to it like `IFuseOptions< T >`.

Sometimes `T` is just `string`, like in the first unit test. And we're just searching through an array of strings. Sometimes (like in the dashboard), `T` will be an object like `{ title: string, domain: string, ... }` since we'll be searching through the array of site data objects. Traditionally generics use short identifiers like `T` or `U` for some reason, but they can be longer. Maybe something like `Needle` would be better? `data: Needle[]` is the entire haystack we're searching, and this hook returns `results: Needle[]` is all of the needles we found using this particular query.

The final thing is `Partial< ... >` which is a built-in TypeScript helper. It makes all of the `IFuseOptions` optional. It probably isn't necessary because all the optional already are optional https://github.com/krisk/Fuse/blob/7ff9c2f2ea8345287b05f7560799c6fee92325e3/src/index.d.ts#L242